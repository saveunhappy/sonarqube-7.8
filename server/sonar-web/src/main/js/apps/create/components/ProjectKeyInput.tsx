/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { debounce } from 'lodash';
import ValidationInput from '../../../components/controls/ValidationInput';
import { doesComponentExists } from '../../../api/components';
import { translate } from '../../../helpers/l10n';

interface Props {
  className?: string;
  value: string;
  onChange: (value: string | undefined) => void;
}

interface State {
  error?: string;
  touched: boolean;
  validating: boolean;
}

export default class ProjectKeyInput extends React.PureComponent<Props, State> {
  mounted = false;
  constructor(props: Props) {
    super(props);
    this.state = { error: undefined, touched: false, validating: false };
    this.checkFreeKey = debounce(this.checkFreeKey, 250);
  }

  componentDidMount() {
    this.mounted = true;
    if (this.props.value) {
      this.validateKey(this.props.value);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  checkFreeKey = (key: string) => {
    this.setState({ validating: true });
    return doesComponentExists({ component: key })
      .then(alreadyExist => {
        if (this.mounted && key === this.props.value) {
          if (!alreadyExist) {
            this.setState({ error: undefined, validating: false });
          } else {
            this.setState({
              error: translate('onboarding.create_project.project_key.taken'),
              touched: true,
              validating: false
            });
          }
        }
      })
      .catch(() => {
        if (this.mounted && key === this.props.value) {
          this.setState({ error: undefined, validating: false });
        }
      });
  };

  handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.currentTarget;
    this.setState({ touched: true });
    this.validateKey(value);
    this.props.onChange(value);
  };

  validateKey(key: string) {
    if (key.length > 400 || !/^[\w-.:]*[a-zA-Z]+[\w-.:]*$/.test(key)) {
      this.setState({
        error: translate('onboarding.create_project.project_key.error'),
        touched: true
      });
    } else {
      this.checkFreeKey(key);
    }
  }

  render() {
    const isInvalid = this.state.touched && this.state.error !== undefined;
    const isValid = this.state.touched && !this.state.validating && this.state.error === undefined;
    return (
      <ValidationInput
        className={this.props.className}
        description={translate('onboarding.create_project.project_key.description')}
        error={this.state.error}
        help={translate('onboarding.create_project.project_key.help')}
        id="project-key"
        isInvalid={isInvalid}
        isValid={isValid}
        label={translate('onboarding.create_project.project_key')}
        required={true}>
        <input
          autoFocus={true}
          className={classNames('input-super-large', {
            'is-invalid': isInvalid,
            'is-valid': isValid
          })}
          id="project-key"
          maxLength={400}
          minLength={1}
          onChange={this.handleChange}
          type="text"
          value={this.props.value}
        />
      </ValidationInput>
    );
  }
}
