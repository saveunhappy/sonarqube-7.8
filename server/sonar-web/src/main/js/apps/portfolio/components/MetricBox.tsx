/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { Link } from 'react-router';
import Effort from './Effort';
import HistoryButtonLink from './HistoryButtonLink';
import MainRating from './MainRating';
import MeasuresButtonLink from './MeasuresButtonLink';
import RatingFreshness from './RatingFreshness';
import { METRICS_PER_TYPE } from '../utils';
import HelpTooltip from '../../../components/controls/HelpTooltip';
import Level from '../../../components/ui/Level';
import Measure from '../../../components/measure/Measure';
import { translate } from '../../../helpers/l10n';
import { getComponentDrilldownUrl } from '../../../helpers/urls';

interface Props {
  component: string;
  measures: T.Dict<string | undefined>;
  metricKey: string;
}

export default function MetricBox({ component, measures, metricKey }: Props) {
  const keys = METRICS_PER_TYPE[metricKey];
  const rating = measures[keys.rating];
  const lastReliabilityChange = measures[keys.last_change];
  const rawEffort = measures[keys.effort];
  const effort = rawEffort ? JSON.parse(rawEffort) : undefined;

  return (
    <div className="portfolio-box">
      <h2 className="portfolio-box-title">
        {translate(keys.label)}
        <HelpTooltip
          className="little-spacer-left"
          overlay={translate('portfolio.metric_domain', metricKey, 'help')}
        />
      </h2>

      {rating ? (
        <MainRating component={component} metric={keys.rating} value={rating} />
      ) : (
        <div className="portfolio-box-rating">
          <span className="rating no-rating">â€”</span>
        </div>
      )}

      {rating && (
        <>
          <h3>{translate('portfolio.metric_trend')}</h3>
          <RatingFreshness lastChange={lastReliabilityChange} rating={rating} />
        </>
      )}

      {metricKey === 'releasability'
        ? Number(effort) > 0 && (
            <>
              <h3>{translate('portfolio.lowest_rated_projects')}</h3>
              <div className="portfolio-effort">
                <Link
                  to={getComponentDrilldownUrl({
                    componentKey: component,
                    metric: 'alert_status'
                  })}>
                  <span>
                    <Measure
                      className="little-spacer-right"
                      metricKey="projects"
                      metricType="SHORT_INT"
                      value={effort}
                    />
                    {Number(effort) === 1
                      ? translate('project_singular')
                      : translate('project_plural')}
                  </span>
                </Link>{' '}
                <Level level="ERROR" small={true} />
              </div>
            </>
          )
        : effort && (
            <>
              <h3>{translate('portfolio.lowest_rated_projects')}</h3>
              <Effort component={component} effort={effort} metricKey={keys.rating} />
            </>
          )}

      <div className="portfolio-box-links">
        <div>
          <MeasuresButtonLink component={component} metric={keys.measuresMetric} />
        </div>
        <div>
          <HistoryButtonLink component={component} metric={keys.activity || keys.rating} />
        </div>
      </div>
    </div>
  );
}
