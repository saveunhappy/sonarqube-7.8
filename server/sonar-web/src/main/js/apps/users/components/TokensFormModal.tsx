/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import TokensForm from './TokensForm';
import Modal from '../../../components/controls/Modal';
import { translate } from '../../../helpers/l10n';
import { ResetButtonLink } from '../../../components/ui/buttons';

interface Props {
  user: T.User;
  onClose: () => void;
  updateTokensCount: (login: string, tokensCount: number) => void;
}

export default function TokensFormModal(props: Props) {
  const header = translate('users.tokens');
  return (
    <Modal contentLabel={header} onRequestClose={props.onClose}>
      <header className="modal-head">
        <h2>{header}</h2>
      </header>
      <div className="modal-body modal-container">
        <TokensForm login={props.user.login} updateTokensCount={props.updateTokensCount} />
      </div>
      <footer className="modal-foot">
        <ResetButtonLink onClick={props.onClose}>{translate('Done')}</ResetButtonLink>
      </footer>
    </Modal>
  );
}
