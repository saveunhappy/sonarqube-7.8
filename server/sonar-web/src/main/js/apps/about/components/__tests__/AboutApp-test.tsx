/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { mount } from 'enzyme';
import { AboutApp } from '../AboutApp';
import { addWhitePageClass, removeWhitePageClass } from '../../../../helpers/pages';
import { searchProjects } from '../../../../api/components';
import { getFacet } from '../../../../api/issues';
import { mockLocation, mockAppState, mockCurrentUser } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';

jest.mock('../../../../components/icons-components/BugIcon');
jest.mock('../../../../components/icons-components/VulnerabilityIcon');
jest.mock('../../../../components/icons-components/CodeSmellIcon');
jest.mock('../../../../components/icons-components/SecurityHotspotIcon');
jest.mock('../../../../components/icons-components/TagsIcon');

jest.mock('../../../../helpers/pages', () => ({
  addWhitePageClass: jest.fn(),
  removeWhitePageClass: jest.fn()
}));

jest.mock('../../../../api/components', () => ({
  searchProjects: jest.fn().mockResolvedValue({ paging: { total: 5 } })
}));

jest.mock('../../../../api/issues', () => ({
  getFacet: jest.fn().mockResolvedValue({
    facet: [
      { count: 5, val: 'CODE_SMELL' },
      { count: 10, val: 'BUG' },
      { count: 0, val: 'VULNERABILITY' },
      { count: 5, val: 'SECURITY_HOTSPOT' }
    ]
  })
}));

jest.mock('../../../../app/components/GlobalContainer', () => ({
  default: class GlobalContainer extends React.Component {
    static displayName = 'GlobalContainer';
    render() {
      return this.props.children;
    }
  }
}));

it('should render correctly', async () => {
  const wrapper = mountRender();
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
  expect(addWhitePageClass).toBeCalled();

  wrapper.unmount();
  expect(removeWhitePageClass).toBeCalled();
});

it('should load issues, projects, and custom text upon mounting', () => {
  const fetchAboutPageSettings = jest.fn();
  mountRender({ fetchAboutPageSettings });
  expect(fetchAboutPageSettings).toBeCalled();
  expect(searchProjects).toBeCalled();
  expect(getFacet).toBeCalled();
});

function mountRender(props: Partial<AboutApp['props']> = {}) {
  return mount(
    <AboutApp
      appState={mockAppState()}
      currentUser={mockCurrentUser()}
      customText="Lorem ipsum"
      fetchAboutPageSettings={jest.fn().mockResolvedValue('')}
      location={mockLocation()}
      {...props}
    />
  );
}
