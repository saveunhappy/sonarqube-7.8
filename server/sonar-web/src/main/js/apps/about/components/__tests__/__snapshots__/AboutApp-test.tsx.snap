// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<AboutApp
  appState={
    Object {
      "defaultOrganization": "foo",
      "edition": "community",
      "productionDatabase": true,
      "qualifiers": Array [
        "TRK",
      ],
      "settings": Object {},
      "version": "1.0",
    }
  }
  currentUser={
    Object {
      "isLoggedIn": false,
    }
  }
  customText="Lorem ipsum"
  fetchAboutPageSettings={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  location={
    Object {
      "action": "PUSH",
      "hash": "",
      "key": "key",
      "pathname": "/path",
      "query": Object {},
      "search": "",
      "state": Object {},
    }
  }
>
  <GlobalContainer
    location={
      Object {
        "action": "PUSH",
        "hash": "",
        "key": "key",
        "pathname": "/path",
        "query": Object {},
        "search": "",
        "state": Object {},
      }
    }
  >
    <div
      className="page page-limited about-page"
      id="about-page"
    >
      <A11ySkipTarget
        anchor="about_main"
      >
        <A11ySkipTargetInner
          addA11ySkipLink={[Function]}
          anchor="about_main"
          removeA11ySkipLink={[Function]}
        >
          <span
            id="a11y_target__about_main"
          />
        </A11ySkipTargetInner>
      </A11ySkipTarget>
      <div
        className="about-page-entry"
      >
        <div
          className="about-page-intro"
        >
          <h1
            className="big-spacer-bottom"
          >
            layout.sonar.slogan
          </h1>
          <Link
            className="button button-active big-spacer-right"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/sessions/new"
          >
            <a
              className="button button-active big-spacer-right"
              onClick={[Function]}
              style={Object {}}
            >
              layout.login
            </a>
          </Link>
          <a
            className="button"
            href="https://redirect.sonarsource.com/doc/home.html"
            rel="noopener noreferrer"
            target="_blank"
          >
            about_page.read_documentation
          </a>
        </div>
        <div
          className="about-page-instance"
        >
          <AboutProjects
            count={5}
            loading={false}
          >
            <div
              className="about-page-projects"
            >
              <div>
                <div>
                  <Link
                    className="about-page-projects-link"
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="/projects"
                  >
                    <a
                      className="about-page-projects-link"
                      onClick={[Function]}
                      style={Object {}}
                    >
                      5
                    </a>
                  </Link>
                </div>
                <div>
                  about_page.projects_analyzed
                </div>
              </div>
            </div>
          </AboutProjects>
          <EntryIssueTypes
            bugs={10}
            codeSmells={5}
            loading={false}
            securityHotspots={5}
            vulnerabilities={0}
          >
            <div
              className="about-page-projects"
            >
              <table
                className="about-page-issue-types"
              >
                <tbody>
                  <tr>
                    <td
                      className="about-page-issue-type-number"
                    >
                      <Link
                        className="about-page-issue-type-link"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/issues",
                            "query": Object {
                              "resolved": "false",
                              "s": "CREATION_DATE",
                              "types": "BUG",
                            },
                          }
                        }
                      >
                        <a
                          className="about-page-issue-type-link"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          10
                        </a>
                      </Link>
                    </td>
                    <td>
                      <span
                        className="little-spacer-right"
                      >
                        <MOCKBugIcon />
                      </span>
                      issue.type.BUG.plural
                    </td>
                  </tr>
                  <tr>
                    <td
                      className="about-page-issue-type-number"
                    >
                      <Link
                        className="about-page-issue-type-link"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/issues",
                            "query": Object {
                              "resolved": "false",
                              "s": "CREATION_DATE",
                              "types": "VULNERABILITY",
                            },
                          }
                        }
                      >
                        <a
                          className="about-page-issue-type-link"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          0
                        </a>
                      </Link>
                    </td>
                    <td>
                      <span
                        className="little-spacer-right"
                      >
                        <MOCKVulnerabilityIcon />
                      </span>
                      issue.type.VULNERABILITY.plural
                    </td>
                  </tr>
                  <tr>
                    <td
                      className="about-page-issue-type-number"
                    >
                      <Link
                        className="about-page-issue-type-link"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/issues",
                            "query": Object {
                              "resolved": "false",
                              "s": "CREATION_DATE",
                              "types": "CODE_SMELL",
                            },
                          }
                        }
                      >
                        <a
                          className="about-page-issue-type-link"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          5
                        </a>
                      </Link>
                    </td>
                    <td>
                      <span
                        className="little-spacer-right"
                      >
                        <MOCKCodeSmellIcon />
                      </span>
                      issue.type.CODE_SMELL.plural
                    </td>
                  </tr>
                  <tr>
                    <td
                      className="about-page-issue-type-number"
                    >
                      <Link
                        className="about-page-issue-type-link"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/issues",
                            "query": Object {
                              "resolved": "false",
                              "s": "CREATION_DATE",
                              "types": "SECURITY_HOTSPOT",
                            },
                          }
                        }
                      >
                        <a
                          className="about-page-issue-type-link"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          5
                        </a>
                      </Link>
                    </td>
                    <td>
                      <span
                        className="little-spacer-right"
                      >
                        <MOCKSecurityHotspotIcon />
                      </span>
                      issue.type.SECURITY_HOTSPOT.plural
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </EntryIssueTypes>
        </div>
      </div>
      <div
        className="about-page-section"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Lorem ipsum",
          }
        }
      />
      <AboutLanguages>
        <div
          className="boxed-group"
        >
          <h2>
            about_page.languages
          </h2>
          <div
            className="boxed-group-inner"
          >
            <p
              className="about-page-text"
            >
              about_page.languages.text
            </p>
            <ul
              className="about-page-languages"
            >
              <li
                key="0"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/java.html"
                >
                  Java
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/vbnet.html"
                >
                  VB.NET
                </a>
              </li>
              <li
                key="1"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/cpp.html"
                >
                  C/C++
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/plsql.html"
                >
                  PL/SQL
                </a>
              </li>
              <li
                key="2"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/csharp.html"
                >
                  C#
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/tsql.html"
                >
                  T-SQL
                </a>
              </li>
              <li
                key="3"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/cobol.html"
                >
                  COBOL
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/flex.html"
                >
                  Flex
                </a>
              </li>
              <li
                key="4"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/abap.html"
                >
                  ABAP
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/python.html"
                >
                  Python
                </a>
              </li>
              <li
                key="5"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/web.html"
                >
                  HTML
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/groovy.html"
                >
                  Groovy
                </a>
              </li>
              <li
                key="6"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/rpg.html"
                >
                  RPG
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/php.html"
                >
                  PHP
                </a>
              </li>
              <li
                key="7"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/javascript.html"
                >
                  JavaScript
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/swift.html"
                >
                  Swift
                </a>
              </li>
              <li
                key="8"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/typescript.html"
                >
                  TypeScript
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/vb.html"
                >
                  Visual Basic
                </a>
              </li>
              <li
                key="9"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/objectivec.html"
                >
                  Objective C
                </a>
                <br />
                <a
                  href="https://redirect.sonarsource.com/plugins/pli.html"
                >
                  PL/I
                </a>
              </li>
              <li
                key="10"
              >
                <a
                  href="https://redirect.sonarsource.com/plugins/xml.html"
                >
                  XML
                </a>
                <br />
              </li>
            </ul>
          </div>
        </div>
      </AboutLanguages>
      <AboutQualityModel>
        <div
          className="boxed-group about-quality-model"
        >
          <h2>
            about_page.quality_model
          </h2>
          <div
            className="boxed-group-inner clearfix"
          >
            <div
              className="flex-columns"
            >
              <div
                className="flex-column flex-column-third"
              >
                <h3>
                  metric_domain.Reliability
                </h3>
                <div
                  className="pull-left little-spacer-right"
                >
                  <MOCKBugIcon />
                </div>
                <p
                  className="about-page-text overflow-hidden"
                >
                  <strong>
                    issue.type.BUG.plural
                  </strong>
                   
                  about_page.quality_model.bugs
                </p>
              </div>
              <div
                className="flex-column flex-column-third"
              >
                <h3>
                  metric_domain.Security
                </h3>
                <div
                  className="pull-left little-spacer-right"
                >
                  <MOCKVulnerabilityIcon />
                </div>
                <p
                  className="about-page-text overflow-hidden"
                >
                  <strong>
                    issue.type.VULNERABILITY.plural
                  </strong>
                   
                  about_page.quality_model.vulnerabilities
                </p>
                <br />
                <div
                  className="pull-left little-spacer-right"
                >
                  <MOCKSecurityHotspotIcon />
                </div>
                <p
                  className="about-page-text overflow-hidden"
                >
                  <strong>
                    issue.type.SECURITY_HOTSPOT.plural
                  </strong>
                   
                  about_page.quality_model.security_hotspots
                </p>
              </div>
              <div
                className="flex-column flex-column-third"
              >
                <h3>
                  metric_domain.Maintainability
                </h3>
                <div
                  className="pull-left little-spacer-right"
                >
                  <MOCKCodeSmellIcon />
                </div>
                <p
                  className="about-page-text overflow-hidden"
                >
                  <strong>
                    issue.type.CODE_SMELL.plural
                  </strong>
                   
                  about_page.quality_model.code_smells
                </p>
              </div>
            </div>
          </div>
        </div>
      </AboutQualityModel>
      <div
        className="flex-columns"
      >
        <div
          className="flex-column flex-column-half about-page-group-boxes"
        >
          <AboutCleanCode>
            <div
              className="boxed-group"
            >
              <h2>
                about_page.clean_code
              </h2>
              <div
                className="boxed-group-inner"
              >
                <p
                  className="about-page-text"
                >
                  about_page.clean_code.text
                </p>
                <ReadMore
                  link="https://redirect.sonarsource.com/doc/issues.html"
                >
                  <div
                    className="big-spacer-top"
                  >
                    <a
                      className="about-page-link-more"
                      href="https://redirect.sonarsource.com/doc/issues.html"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span>
                        about_page.read_more
                      </span>
                    </a>
                  </div>
                </ReadMore>
              </div>
            </div>
          </AboutCleanCode>
        </div>
        <div
          className="flex-column flex-column-half about-page-group-boxes"
        >
          <AboutLeakPeriod>
            <div
              className="boxed-group"
            >
              <h2>
                about_page.fix_the_leak
              </h2>
              <div
                className="boxed-group-inner"
              >
                <p
                  className="about-page-text"
                >
                  about_page.fix_the_leak_on_new_code.text
                </p>
                <ReadMore
                  link="https://redirect.sonarsource.com/doc/fix-the-leak.html"
                >
                  <div
                    className="big-spacer-top"
                  >
                    <a
                      className="about-page-link-more"
                      href="https://redirect.sonarsource.com/doc/fix-the-leak.html"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span>
                        about_page.read_more
                      </span>
                    </a>
                  </div>
                </ReadMore>
              </div>
            </div>
          </AboutLeakPeriod>
        </div>
      </div>
      <div
        className="flex-columns"
      >
        <div
          className="flex-column flex-column-half about-page-group-boxes"
        >
          <AboutQualityGates>
            <div
              className="boxed-group"
            >
              <h2>
                about_page.quality_gates
              </h2>
              <div
                className="boxed-group-inner"
              >
                <p
                  className="about-page-text"
                >
                  about_page.quality_gates.text
                </p>
                <ReadMore
                  link="https://redirect.sonarsource.com/doc/quality-gates.html"
                >
                  <div
                    className="big-spacer-top"
                  >
                    <a
                      className="about-page-link-more"
                      href="https://redirect.sonarsource.com/doc/quality-gates.html"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span>
                        about_page.read_more
                      </span>
                    </a>
                  </div>
                </ReadMore>
              </div>
            </div>
          </AboutQualityGates>
        </div>
        <div
          className="flex-column flex-column-half about-page-group-boxes"
        >
          <AboutStandards
            appState={
              Object {
                "defaultOrganization": "foo",
                "edition": "community",
                "productionDatabase": true,
                "qualifiers": Array [
                  "TRK",
                ],
                "settings": Object {},
                "version": "1.0",
              }
            }
          >
            <div
              className="boxed-group"
            >
              <h2>
                about_page.standards
              </h2>
              <div
                className="boxed-group-inner"
              >
                <p
                  className="about-page-text"
                >
                  about_page.standards.text
                </p>
                <div
                  className="spacer-top"
                >
                  <ul
                    className="list-inline"
                  >
                    <li>
                      <Link
                        className="link-with-icon"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/coding_rules",
                            "query": Object {
                              "tags": "misra",
                            },
                          }
                        }
                      >
                        <a
                          className="link-with-icon"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <MOCKTagsIcon />
                          <span
                            className="little-spacer-left"
                          >
                            MISRA
                          </span>
                        </a>
                      </Link>
                    </li>
                    <li>
                      <Link
                        className="link-with-icon"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/coding_rules",
                            "query": Object {
                              "tags": "cert",
                            },
                          }
                        }
                      >
                        <a
                          className="link-with-icon"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <MOCKTagsIcon />
                          <span
                            className="little-spacer-left"
                          >
                            CERT
                          </span>
                        </a>
                      </Link>
                    </li>
                    <li>
                      <Link
                        className="link-with-icon"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/coding_rules",
                            "query": Object {
                              "tags": "cwe",
                            },
                          }
                        }
                      >
                        <a
                          className="link-with-icon"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <MOCKTagsIcon />
                          <span
                            className="little-spacer-left"
                          >
                            CWE
                          </span>
                        </a>
                      </Link>
                    </li>
                    <li>
                      <Link
                        className="link-with-icon"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/coding_rules",
                            "query": Object {
                              "tags": "owasp-a1,owasp-a2,owasp-a3,owasp-a4,owasp-a5,owasp-a6,owasp-a7,owasp-a8,owasp-a9,owasp-a10",
                            },
                          }
                        }
                      >
                        <a
                          className="link-with-icon"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <MOCKTagsIcon />
                          <span
                            className="little-spacer-left"
                          >
                            OWASP Top 10
                          </span>
                        </a>
                      </Link>
                    </li>
                    <li>
                      <Link
                        className="link-with-icon"
                        onlyActiveOnIndex={false}
                        style={Object {}}
                        to={
                          Object {
                            "pathname": "/coding_rules",
                            "query": Object {
                              "tags": "sans-top25-porous,sans-top25-risky,sans-top25-insecure",
                            },
                          }
                        }
                      >
                        <a
                          className="link-with-icon"
                          onClick={[Function]}
                          style={Object {}}
                        >
                          <MOCKTagsIcon />
                          <span
                            className="little-spacer-left"
                          >
                            SANS Top 25
                          </span>
                        </a>
                      </Link>
                    </li>
                  </ul>
                </div>
                <ReadMore
                  link="https://redirect.sonarsource.com/doc/rules.html"
                >
                  <div
                    className="big-spacer-top"
                  >
                    <a
                      className="about-page-link-more"
                      href="https://redirect.sonarsource.com/doc/rules.html"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      <span>
                        about_page.read_more
                      </span>
                    </a>
                  </div>
                </ReadMore>
              </div>
            </div>
          </AboutStandards>
        </div>
      </div>
      <AboutScanners>
        <div
          className="boxed-group"
        >
          <h2>
            about_page.scanners
          </h2>
          <div
            className="boxed-group-inner"
          >
            <p
              className="about-page-text"
            >
              about_page.scanners.text
            </p>
            <div
              className="about-page-analyzers"
            >
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/doc/install-configure-scanner.html"
                key="sonarqube"
              >
                <img
                  alt="about_page.scanners.sonarqube"
                  height={60}
                  src="/images/scanner-logos/sonarqube.svg"
                />
              </a>
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/doc/install-configure-scanner-msbuild.html"
                key="msbuild"
              >
                <img
                  alt="about_page.scanners.msbuild"
                  height={60}
                  src="/images/scanner-logos/msbuild.svg"
                />
              </a>
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/doc/install-configure-scanner-maven.html"
                key="maven"
              >
                <img
                  alt="about_page.scanners.maven"
                  height={60}
                  src="/images/scanner-logos/maven.svg"
                />
              </a>
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/doc/gradle.html"
                key="gradle"
              >
                <img
                  alt="about_page.scanners.gradle"
                  height={60}
                  src="/images/scanner-logos/gradle.svg"
                />
              </a>
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/plugins/jenkins.html"
                key="jenkins"
              >
                <img
                  alt="about_page.scanners.jenkins"
                  height={60}
                  src="/images/scanner-logos/jenkins.svg"
                />
              </a>
              <a
                className="about-page-analyzer-box"
                href="https://redirect.sonarsource.com/doc/install-configure-scanner-ant.html"
                key="ant"
              >
                <img
                  alt="about_page.scanners.ant"
                  height={60}
                  src="/images/scanner-logos/ant.svg"
                />
              </a>
            </div>
          </div>
        </div>
      </AboutScanners>
    </div>
  </GlobalContainer>
</AboutApp>
`;
